src/server.rs

```diff
@@ -1,63 +1,51 @@
-use Application;
+use std::net::SocketAddr;
 
-use std::io;
-use std::net::*;
-use std::ops::DerefMut;
-use std::sync::{Arc, Mutex};
-use std::thread;
+use tokio;
+use tokio::codec::Decoder;
+use tokio::io;
+use tokio::net::TcpListener;
+use tokio::prelude::*;
 
+use codec::abci::ABCICodec;
 use messages::abci::*;
-use stream::AbciStream;
+use Application;
 
 /// Creates the TCP server and listens for connections from Tendermint
 pub fn serve<A>(app: A, addr: SocketAddr) -> io::Result<()>
 where
-    A: Application + 'static + Send + Sync,
+    A: Application + 'static + Send + Sync + Copy,
 {
-    let listener = TcpListener::bind(addr).unwrap();
+    let listener = TcpListener::bind(&addr).unwrap();
+    let incoming = listener.incoming();
+    let server = incoming
+        .map_err(|err| println!("Connection failed: {}", err))
+        .for_each(move |socket| {
+            println!("Got connection! {:?}", socket);
+            let framed = ABCICodec::new().framed(socket);
+            let (_writer, reader) = framed.split();
 
-    // Wrap the app atomically and clone for each connection.
-    let app = Arc::new(Mutex::new(app));
```

Comment:
Not sure if this can be removed (and `Copy` can't really be required, as mentioned in https://github.com/tendermint/rust-abci/pull/50/files#diff-b4aea3e418ccdb71239b96952d9cddb6R108 ) . When I mentioned in https://github.com/tendermint/rust-abci/issues/31#issue-386110209 that there may not be a reason for mutex, I meant the following possibilities:

* if the request reading / parsing is separate from response generation, the response generation could potentially be in one system thread, so no need for app synchronisation / locking... the app could then say be wrapped say in `Rc<Cell<...>>` or owned within that "response generating" thread's scope
* if the `Application` trait is changed, such that it doesn't require a mutable ref in Mempool/Info connection-related operations (CheckTx, Query...), there could be a `RwLock` instead of `Mutex`... this could be good, but could also lead to some problems, e.g. a writer starvation.
* if the `Application` trait is split into some sort of "Query" and "Update" parts and it's somehow possible to distinguish among those 3 ABCI connections, it may be possible to use them without dynamic wrapping -- Mempool and Info connections would get "&ApplicationQuery", Consensus connection would get "&mut ApplicationUpdate".

Perhaps the first possibility can be easily applied here, but if not, I'd leave it for now and open a separate issue for this, because it requires a significant re-architecting of rust-abci.
@davebryson any thoughts?